jcl/src/java.base/share/classes/openj9/internal/foreign/abi/LayoutStrPreprocessor.java

```diff
@@ -209,38 +215,151 @@ private static StringBuilder preprocessLayout(MemoryLayout targetLayout, boolean
+			break;
+		case 8: /* 8 Bytes. */
+			if (validatePrimTypesOfUnionForGPR(unionLayout, byteAlignment)) {
+				sequenceElement = JAVA_LONG;
+			} else {
+				sequenceElement = JAVA_DOUBLE;
+			}
+			break;
+		default:
+			throw new IllegalArgumentException("Invalid alignment of the union layout: " + unionLayout);
 		}
 
-		return targetLayoutString;
+		long elementByteSize = sequenceElement.byteSize();
+		long sequenceCount = (0 == (unionByteSize % elementByteSize)) ?
+				(unionByteSize / elementByteSize) : (unionByteSize / elementByteSize + 1);
+
+		/* There is no need to create a sequence layout for a single element. */
+		MemoryLayout sequenceLayout = (sequenceCount > 1) ?
+				MemoryLayout.sequenceLayout(sequenceCount, sequenceElement) : sequenceElement;
+		elementLayoutStrs.append(preprocessLayout(sequenceLayout, isDownCall));
+
+		/* Prefix "#" to denote the start of this layout string in the case of downcall. */
+		if (isDownCall) {
+			targetLayoutStr.append('#').append(1);
+		}
+		targetLayoutStr.append('[').append(elementLayoutStrs).append(']');
+
+		return targetLayoutStr;
 	}
 
+	/* Check whether the specified 4/8-byte aligned group layout (struct/union) layout contains
+	 * non-floating-point primitves so as to determine whether the elements of union can be safely
+	 * mapped to GPRs.
+	 */
+	private static boolean validatePrimTypesOfUnionForGPR(GroupLayout groupLayout, int byteAlignment) {
+		List<MemoryLayout> elementLayoutList = groupLayout.memberLayouts();
+		boolean arePrimTypesForGPR = false;
+		for (MemoryLayout elementLayout : elementLayoutList) {
+			if (((elementLayout instanceof ValueLayout elemValueLayout) && (validatePrimTypeForGPR(elemValueLayout, byteAlignment)))
+				|| ((elementLayout instanceof GroupLayout elemGroupLayout) && validatePrimTypesOfUnionForGPR(elemGroupLayout, byteAlignment))
+			) {
+				arePrimTypesForGPR = true;
+				break;
+			}
+		}
+
+		return arePrimTypesForGPR;
+	}
+
+	/* Check whether the primitive type can be safely mappped to GPR when it is not float/double
+	 * in the struct/union including some exceptions due to the platform restrictions.
+	 *
+	 * Note:
+	 * The primitive type is mappped to GPR as long as it is not float/double on most platforms
+	 * with a couple of exceptions that:
+	 * 1) the biggest alignment is 8 bytes on the non-X86_64 platforms even though the type is float.
+	 * 2) it is always mapped onto GPR in any case on zLinux.
+	 */
+	private static boolean validatePrimTypeForGPR(ValueLayout primLayout, int byteAlignment) {
+		Class<?> carrier = primLayout.carrier();
+		boolean isX86_64 = arch.equals("amd64") || arch.equals("x86_64");
+		boolean isLinuxS390x = arch.equals("s390x") && osName.startsWith("linux");
+		boolean isPrimTypeForGPR = false;
+
+		if (((carrier != double.class) && (carrier != float.class))
+			|| ((carrier == float.class) && (8 == byteAlignment) && !isX86_64)
+			|| isLinuxS390x
+		) {
+			isPrimTypeForGPR = true;
+		}
```

Comment:
Using suggested fields and avoiding carrier tests:
```java
		boolean isPrimTypeForGPR;

		if (carrier == float.class) {
			isPrimTypeForGPR = isLinuxS390x || ((8 == byteAlignment) && !isX86_64);
		} else if (carrier == double.class) {
			isPrimTypeForGPR = false;
		} else {
			isPrimTypeForGPR = true;
		}
```

Reply:
Fixed as suggested above.

As for:
```
} else if (carrier == double.class) {
			isPrimTypeForGPR = false; <-----
```
it should be `isPrimTypeForGPR = isLinuxS390x;` as any check on zLinux is ignored.